### PR TITLE
ENH: Adds weight reg alpha reduction with batches

### PR DESCRIPTION
**Pull request by jfkirk**
Addresses https://github.com/jfkirk/tensorrec/issues/30

Also adds a nice convenience script to run through Movielens with all the loss functions

### CODE REVIEW

File: tensorrec/tensorrec.py [hunk 1]
```diff
@@ -405,6 +405,8 @@ def fit_partial(self, interactions, user_features, item_features, epochs=1, lear
         if self.loss_graph.is_sample_based:
             if (n_sampled_items is None) or (n_sampled_items <= 0):
                 raise ValueError("n_sampled_items must be an integer >0")
+        if (n_sampled_items is not None) and (not self.loss_graph.is_sample_based):
+            logging.warning('n_sampled_items was specified, but the loss graph is not sample-based')
```
Comment (jcauteru) [inline tensorrec/tensorrec.py]: need to think about the math but we might want to warn if `n_sampled_items == 1 `...degenerate case?

File: tensorrec/tensorrec.py [hunk 2]
```diff
@@ -421,8 +423,12 @@ def fit_partial(self, interactions, user_features, item_features, epochs=1, lear
                                                            user_features=user_features,
                                                            item_features=item_features,
                                                            user_batch_size=user_batch_size,
-                                                           extra_feed_kwargs={self.tf_learning_rate: learning_rate,
-                                                                              self.tf_alpha: alpha})
+                                                           extra_feed_kwargs={self.tf_learning_rate: learning_rate})
+
+        # This scales down the alpha based on the number of batches and inserts the new alpha in all feed dicts
+        batched_alpha = calculate_batched_alpha(num_batches=len(batched_feed_dicts), alpha=alpha)
+        for feed_dict in batched_feed_dicts:
```
Comment (jcauteru) [inline tensorrec/tensorrec.py]: Only issue I see here, maybe an edge case (or not even possible to change) but (i think) this fixes the order of batched_feed_dicts for training. Any concern that someone somehow might mess with the batch order?

File: tensorrec/tensorrec.py (hunk 1 above)
Comment (jfkirk) [reply tensorrec/tensorrec.py]: I think `n_sampled_items == 1` isn't good practice but isn't necessarily degenerate. If we warn on that, we're saying "1 is too low of a value" but I'm hesitant to define "too low" for all problems. 

For example: Is 2 too low for all problems? Probably... but I don't think we'd warn on that, either.

What do you think?

File: tensorrec/tensorrec.py (hunk 2 above)
Comment (jfkirk) [reply tensorrec/tensorrec.py]: I think messing with batch order would be okay. In fact, we may want to add `shuffle_batches` as a fit arg to intentionally shuffle around the batches every epoch.

File: tensorrec/tensorrec.py (hunk 1 above)
Comment (jcauteru) [reply tensorrec/tensorrec.py]: yea, I guess degenerate is to strong. unexpected behavior...but yea I guess we can't really say anything on the "correct" n-sample

File: tensorrec/tensorrec.py (hunk 2 above)
Comment (jcauteru) [reply tensorrec/tensorrec.py]: :+1:

**Comment by jcauteru** [comment]
:shipit:
